### DISCUSSION TITLE
Pre-Cache all routes and make available in offline

### DISCUSSION THREAD

**Discussion by asif-selise** [serwist/serwist#200] [Q&A]
I have a Next 14 project and I'm using app route in it. I also have dynamic routes in it. I'm using roouter.push() from next/navigation.

In my home page I'm calling multiple API's and storing the fetched data in indexDB in browser. On my other pages/routes, I'm manipulating these data in indexDB. whenever user comes back to the home page the updated data in indexDB gets synced with API. This is how it is working. 

Now I want my application to work in offline.

I've implemented service worker hence PWA in it, following recent Next.js's recently released doc ([link](https://nextjs.org/docs/app/building-your-application/configuring/progressive-web-apps)). There they suggest 'Serwist' to get offline support.

I've installed serwist but the problem is, it only caches the routes that has been visited while online and falls back to the fallback page for unvisited routes. But I want all of my routes to be available in offline. It need to pre-cache all my routes at the beginning.

Here my file/code structure :

**_app/manifest.ts :_**

```typescript
import type { MetadataRoute } from 'next';

export default function manifest(): MetadataRoute.Manifest {
  return {
    name: 'My app',
    icons: [
      {
        purpose: 'maskable',
        sizes: '512x512',
        src: '/assets/images/logo_maskable_pwa.png',
        type: 'image/png',
      },
      {
        purpose: 'any',
        sizes: '512x512',
        src: '/assets/images/logo_rounded_pwa.png',
        type: 'image/png',
      },
    ],
    start_url: '/',
    orientation: 'any',
    display: 'standalone',
    theme_color: '#0056A4',
    background_color: '#FFFFFF',
  };
}
```

**_app/sw.ts :_**

```javascript
import { defaultCache } from '@serwist/next/worker';
import type { PrecacheEntry, SerwistGlobalConfig } from 'serwist';
import { Serwist } from 'serwist';

declare global {
  interface WorkerGlobalScope extends SerwistGlobalConfig {
    __SW_MANIFEST: Array<PrecacheEntry | string> | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
  fallbacks: {
    entries: [
      {
        url: '/~offline',
        matcher({ request }) {
          return request.destination === 'document';
        },
      },
    ],
  },
});

serwist.addEventListeners();
```

**_next.config.mjs :_**

```javascript
import withSerwistInit from '@serwist/next';
const revision = crypto.randomUUID();

const withSerwist = withSerwistInit({
  cacheOnNavigation: true,
  swSrc: 'app/sw.ts',
  swDest: 'public/sw.js',
  additionalPrecacheEntries: [{ url: '/~offline', revision }],
  reloadOnOnline: true,
});

/** @type {import("next").NextConfig} */
const nextConfig = {
  reactStrictMode: true,
};

export default withSerwist(nextConfig);
```

**Comment by DuCanhGH** [comment]
@asif-selise 

You can use `additionalPrecacheEntries`, as shown in your current next.config.js. I'd use a more stable `revision` than `crypto.randomUUID()` though.

Admittedly, this feels painful, but there isn't really a reliable way for Serwist to detect the pages. With how Next.js is structured, `@serwist/next` is forced to rely on `@serwist/webpack-plugin` and run after Next.js bundles your application, not after it prerenders your pages like what `@serwist/vite` can do.

I'd love for Next.js to be more friendly to plugins, to be honest. Still, I guess right now you can write an utility function that hashes your pages' content into `revision`. This isn't reliable, of course, since the modules that your pages import may change, but your pages may not, but I guess it may at least work.

One other option is... cache as your service worker is being installed:

```ts
const serwist = new Serwist(options);

const urlsToPrecache = ["/", "/foo", "/bar"] as const;

self.addEventListener("install", (event) => {
  const requestPromises = Promise.all(
    urlsToPrecache.map((entry) => {
      return serwist.handleRequest({ request: new Request(entry), event });
    }),
  );

  event.waitUntil(requestPromises);
});

serwist.addEventListeners();
```

This works almost like precaching, but the pages are reliably updated every time you update your service worker. Note that this is thanks to the fact that it always updates all the entries rather than uses `revision` to determine if an entry should be updated.

Also this method allows the pages to use the same strategy that they'd match if they weren't precached, unlike when you use `additionalPrecacheEntries`, in which case your pages are forced to use `PrecacheStrategy`.

**Reply by asif-selise** [reply]
Thanks for your response, I'll try this approaches  and let you know the update
